packages/zoe/src/contracts/operaConcertTicket.js

```diff
@@ -0,0 +1,120 @@
+
+  const zoeService = zoe.getZoeService();
+
+  // Create the internal ticket mint
+  const { issuer, mint, amountMath } = produceIssuer('Opera tickets', 'set');
+  await zoe.addNewIssuer(issuer, 'Auditorium');
+
+  // create Zoe helpers after zoe.addNewIssuer because of https://github.com/Agoric/agoric-sdk/issues/802
+  const {
+    rejectOffer,
+    checkIfProposal,
+    swap,
+    canTradeWith,
+    getActiveOffers,
+    assertKeywords,
+  } = makeZoeHelpers(zoe);
+  
+  const { 
+    terms: { show, start, count, expectedAmountPerTicket }, 
+    issuerKeywordRecord: {Buyer: buyerIssuer}
+  } = zoe.getInstanceRecord();
+  
+  const availableTicketDescriptionByNumber = new Map(Array(count)
+    .fill()
+    .map((_, i) => [i+1, harden({ show, start, number: i + 1 })]))
+  
+  const salesPayouts = new Set()
+
+  const auditoriumSeat = harden({
+    async getSalesPayment(){
+      const salesPaymentsP = Promise.all([...salesPayouts])
+        .then(payouts => Promise.all(payouts.map(payout => payout.Buyer)))
+      
+      return salesPaymentsP.then(salesPayments => buyerIssuer.combine(salesPayments));
+    }
+  })
+  const auditoriumInvite = zoe.makeInvite(auditoriumSeat);
+
+  const makeBuyerInvite = () => {
+    const seat = harden({
+      performExchange: async () => {
+        const buyerInviteHandle = inviteHandle;
+        const buyerOffer = zoe.getOffer(buyerInviteHandle)
+        const buyerWant = buyerOffer.proposal.want.Auditorium;
+
+        const ticketNumbers = buyerWant.extent.map(e => e.number)
+
+        const unavailableTicketNumbers = ticketNumbers.filter(number => (!availableTicketDescriptionByNumber.has(number)))
+
+        if(unavailableTicketNumbers.length >= 1){
+          return rejectOffer(inviteHandle, `Some tickets (${unavailableTicketNumbers.join(', ')}) are not available anymore`)
+        }
+
+        const ticketsAmount = amountMath.make(
+          harden(ticketNumbers.map(number => availableTicketDescriptionByNumber.get(number)))
+        )
+
+        // create an zoe invite...
+        const {invite, inviteHandle: ticketInviteHandle} = zoe.makeInvite()
+        // ...and redeem it right away
+        const {payout} = await zoeService.redeem(
+          invite,
+          harden({
+            want: {Buyer: buyerIssuer.getAmountMath().make(expectedAmountPerTicket.extent*ticketNumbers.length)}, 
+            give: {Auditorium: ticketsAmount}
+          }),
+          harden({Auditorium: mint.mintPayment(ticketsAmount)}) // mint and pass to Zoe right away
+        )
+
+        salesPayouts.add(payout)
+
+        try{
+          swap(ticketInviteHandle, buyerInviteHandle)
+        }
+        catch(e){
+          throw e;
+        }
+        // swap was successful
+        for(const number of ticketNumbers)
+          availableTicketDescriptionByNumber.delete(number)
+      },
+    });
+    const { invite, inviteHandle } = zoe.makeInvite(seat);
+    return { invite, inviteHandle };
```

Comment:
I think we should only return the invite here. It should enable more promise pipelining.